pkg/internal/parser.go

```diff
@@ -162,37 +152,280 @@ func sanitize(groupName string) string {
 	return groupName
 }
 
-func (p *packages) WriteContent() error {
+// TODO: Clean this up. Its not required to convert the input to a struct
+// and then to a map before passing it to template. Using of map is better,
+// as it allows to add more variables dynamically.
+func (p *packages) WriteContent(importList *[]string) error {
+	if importList != nil {
+		p.Imports = append(p.Imports, p.appendGenericImports(*importList)...)
+	}
 	templ, err := template.New("client").Parse(commonTempl)
 	if err != nil {
 		return err
 	}
 	return templ.Execute(p.writer, p)
 }
 
-func NewAPI(root *loader.Package, info *markers.TypeInfo, version, group string, isNamespaced bool, hasStatus bool, w io.Writer) (*api, error) {
+// api contains info about each type
+type api struct {
+	Name                      string
+	PkgName                   string
+	Version                   string
+	IsNamespaced              bool
+	AdditionalMethods         []AdditionalMethod
+	SkipVerbs                 []string
+	NoVerbs                   bool
+	HasStatus                 bool
+	ApplyConfigurationPackage string
+	writer                    io.Writer
+	importList                *[]string
+
+	PkgNameUpperFirst string
+	VersionUpperFirst string
+	NameLowerFirst    string
```

Comment:
You can expose formatting methods to be used directly in the Go template ala https://github.com/njhale/code-generator/blob/77c28e0b5b75e66b76ec77e7ca1379871c96bdeb/pkg/internal/informergen/parser.go#L73 and https://github.com/njhale/code-generator/blob/77c28e0b5b75e66b76ec77e7ca1379871c96bdeb/pkg/internal/informergen/templates.go#L39

Reply:
Yeah, that's what I was planning to do here in a follow-up: https://github.com/kcp-dev/code-generator/commit/bf72d7fe23d6d57551e11b7b197c1eecaabf7c9f#diff-7cd18745c2d89da8fe056843208f0d785327606243491e7044af3bd0e4dedaf1R387-R395. Removing all the formatting from structs would clear up the code.